### PR TITLE
 update of base perturbator and inference wrapper to avoid useless embeddings generation during perturbation

### PR DESCRIPTION
**Pull request by thomas-mullor**
## Description

- update of base perturbator and inference wrapper to avoid useless embeddings generation during perturbation
- test of mask application for mask based perturbators

## Related Issue

Related to #2 (see if more tests are needed

- 🥂 Improvement (non-breaking change which improves an existing feature)

## Checklist

- [x] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/FOR-sight-ai/interpreto/blob/main/CODE_OF_CONDUCT.md) document.
- [x] I've read the [`CONTRIBUTING.md`](https://github.com/FOR-sight-ai/interpreto/blob/main/docs/contributing.md) guide.
- [x] I've successfully run the style checks using `make lint`.
- [x] I've written tests for all new methods and classes that I created and successfully ran `make test`.
- [x] I've written the docstring in Google format for all the methods and classes that I used.

### CODE REVIEW

File: interpreto/commons/granularity.py [hunk 1]
```diff
@@ -419,7 +403,7 @@ def __spacy_get_indices(input_ids, offsets, tokenizer, level) -> list[list[int]]
     def aggregate_score_for_gradient_method(
         contribution: torch.Tensor,
         granularity: Granularity | None,
-        granularity_aggregation_strategy: AggregationProtocol | None = None,
+        granularity_aggregation_strategy: Callable[[torch.Tensor, int], torch.Tensor] | None = None,
```
Comment (AntoninPoche) [inline interpreto/commons/granularity.py]: If I remember correctly, we have an Enum now. Could we use it for typing?

File: docs/notebooks/attribution_walkthrough.ipynb
Comment (AntoninPoche) [inline docs/notebooks/attribution_walkthrough.ipynb:1]: If I remember correctly, we used `Granularity.aggregation_strategies` in this notebook, which is now obsolete.

If so, the notebook should be updated accordingly.

File: interpreto/attributions/perturbations/base.py [hunk 2]
```diff
@@ -113,86 +204,38 @@ def _embed(self, model_inputs: TensorMapping) -> TensorMapping:
...
-            return (inputs, mask)
-
-    @jaxtyped(typechecker=beartype)
-    def perturb_ids(self, model_inputs: TensorMapping) -> tuple[TensorMapping, Float[torch.Tensor, "p g"] | None]:
-        """
-        Perturb the input of the model
-
-        Args:
-            model_inputs (MutableMapping): Mapping given by the tokenizer
+class MaskBasedIdsPerturbator(IdsPerturbator):
```
Comment (AntoninPoche) [inline interpreto/attributions/perturbations/base.py]: I do not think there is any `IdsPerturbator` that is not based on a mask.

Do you think we can merge the two into an `IdsPerturbator`?

File: interpreto/attributions/perturbations/base.py [hunk 3]
```diff
@@ -32,8 +32,8 @@
 
 import torch
 from beartype import beartype
-from jaxtyping import Float, Int, jaxtyped
-from transformers import PreTrainedTokenizer
+from jaxtyping import Float, Real, jaxtyped
```
Comment (AntoninPoche) [inline interpreto/attributions/perturbations/base.py]: Any reason to prefer `Real` over `Float` and `Int`?

I made the difference initially because you cannot multiply an int by a float. Thus, it allows me to know in advance, but I might be wrong about compatibility.

**Review by AntoninPoche — CHANGES_REQUESTED** [review]
Nicely done @thomas-mullor !!

I made a few comments on the files, but nothing of importance.

I have two additional general remarks:
- The tests do not pass. I did not look into it, but we can try if you want.
- The split between `IdsPerturbator` and `EmbeddingsPerturbator` partially solves issue #64 (it's the main point). However, some points are not addressed; solving all points of an issue might make sense. (The others are minor modifications).

File: interpreto/attributions/perturbations/base.py [hunk 4]
```diff
@@ -107,94 +174,50 @@ def _embed(self, model_inputs: TensorMapping) -> TensorMapping:
...
-    This class should not be subclasses by perturbation methods.
-    Please consider using TokenMaskBasedPerturbator or EmbeddingsMaskBasedPerturbator instead, depending on where you want to apply your mask
-    """
+        # granularity level of the perturbation (token masking, word masking...)
+        # in most commons cases, this should be set to Granularity.TOKEN
+        self.granularity = granularity
 
-    __slots__ = ()
+    def perturb(self, model_inputs: TensorMapping) -> tuple[TensorMapping, torch.Tensor | None]:
+        return self.perturb_ids(model_inputs)
```
Comment (AntoninPoche) [inline interpreto/attributions/perturbations/base.py]: Why not just rename the `perturb_ids` `perturb`?

File: interpreto/commons/granularity.py [hunk 5]
```diff
@@ -164,7 +148,6 @@ class Granularity:
     SENTENCE = "sentence"  # Sentences of the input
     # PARAGRAPH = "paragraph"  # Not supported yet, the "\n\n" characters are replaced by spaces in many tokenizers.
     DEFAULT = ALL_TOKENS
-    aggregation_strategies = GranularityAggregationStrategy
```
Comment (AntoninPoche) [inline interpreto/commons/granularity.py:167]: Was it not used anywhere in the code?

Comment (thomas-mullor) [reply interpreto/commons/granularity.py:167]: no

**Review by AntoninPoche — APPROVED** [review]
Nice work!

I will manage the version bump and merge.
